src/main/java/org/elasticsearch/search/MultiValueMode.java

```diff
@@ -451,25 +462,33 @@ public long get(int rootDoc) {
                     return missingValue;
                 }
 
-                final int prevRootDoc = rootDocs.prevSetBit(rootDoc - 1);
-                final int firstNestedDoc = innerDocs.nextSetBit(prevRootDoc + 1);
-
-                long accumulated = startLong();
-                int numValues = 0;
-
-                for (int doc = firstNestedDoc; doc != -1 && doc < rootDoc; doc = innerDocs.nextSetBit(doc + 1)) {
-                    values.setDocument(doc);
-                    final int count = values.count();
-                    for (int i = 0; i < count; ++i) {
-                        final long value = values.valueAt(i);
-                        accumulated = apply(accumulated, value);
+                // If via compareBottom this method has previously invoked for the same rootDoc then we need to use the
+                // last seen value as innerDocs can't re-iterate over nested child docs it already emitted,
+                // because DocIdSetIterator can only advance forwards.
+                if (rootDoc == lastSeenRootDoc) {
+                    return lastEmittedValue;
+                }
+                try {
+                    final int prevRootDoc = rootDocs.prevSetBit(rootDoc - 1);
+                    final int firstNestedDoc = innerDocs.advance(prevRootDoc + 1);
+
+                    long accumulated = startLong();
+                    int numValues = 0;
+
+                    for (int doc = firstNestedDoc; doc > prevRootDoc && doc < rootDoc; doc = innerDocs.advance(doc + 1)) {
+                        values.setDocument(doc);
+                        final int count = values.count();
+                        for (int i = 0; i < count; ++i) {
+                            final long value = values.valueAt(i);
+                            accumulated = apply(accumulated, value);
+                        }
+                        numValues += count;
                     }
-                    numValues += count;
+                    lastSeenRootDoc = rootDoc;
+                    return lastEmittedValue = numValues == 0 ? missingValue : reduce(accumulated, numValues);
```

Comment:
this ternary statement is a bit long, maybe a regular if/else would make it easier to read?